### PR TITLE
hdfs: initial implementation #42

### PR DESCRIPTION
**Pull request by urykhy**
#### What is the purpose of this change?

add hdfs support

#### Was the change discussed in an issue or in the forum before?

https://github.com/rclone/rclone/issues/42

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: docs/content/hdfs.md
Comment (ncw) [inline docs/content/hdfs.md:111]: Could we set the environment variable from the rclone config?

**Review by ncw — COMMENTED** [review]
I think this is looking really good - well done!

Next I'd like some way of running it myself so I can put it into the integration tests.

Is there a docker file with a test HDFS available do you know? That would be perfect.

For example this code https://github.com/rclone/rclone/blob/master/fstest/testserver/init.d/TestSwiftAIO launches a test OpenStack Swift storage backend so the integration tests can run against it.

File: docs/content/hdfs.md
Comment (urykhy) [reply docs/content/hdfs.md]: done in new commit

Comment (ncw) [reply docs/content/hdfs.md]: Great

**Comment by ncw** [comment]
Any thoughts on this bit?

> Next I'd like some way of running it myself so I can put it into the integration tests.
> 
> Is there a docker file with a test HDFS available do you know? That would be perfect.
> 
> For example this code https://github.com/rclone/rclone/blob/master/fstest/testserver/init.d/TestSwiftAIO launches a test OpenStack Swift storage backend so the integration tests can run against it.

**Comment by urykhy** [comment]
i add a commit with hdfs docker image, is it works not only for me ? ;)

**Comment by ei-grad** [comment]
Great work! Thanks! :)

Tried this on our production hdfs with `rclone ncdu` - for the whole warehouse it exits silently with code 1 after some time during the process of traversing all the filesystem :-(. I guess it may be related to access denied errors appearing.

On a small subtree of HDFS, where there are no access denied errors, it works like a charm!

**Comment by ivandeex** [comment]
@ei-grad 
Let's help @urykhy understand the problem.
Please run `rclone ncdu yourhdfs: -vvv --log-file test1.log`, redact any secrets (replace by _xxx_) if needed, and send here the log (dragging a file into github comment in the browser will attach it).

**Comment by ivandeex** [comment]
Though the case reported by @ei-grad might point to a bug in the new backend, there exist a known _ncdu_ issue https://github.com/rclone/rclone/issues/4014 and its unmerged fix https://github.com/rclone/rclone/pull/4168, so using `rclone lsl -vvv hdfs:` looks like a better test method.

**Comment by ivandeex** [comment]
@urykhy 
Do the integration pass for you?

**Comment by urykhy** [comment]
@ivandeex 
yes.

```
cd backend/hdfs
GO111MODULE=on go test -v
=== RUN   TestIntegration
    fstests.go:357: Using remote "TestHdfs:"
2020/11/24 19:35:09 NOTICE: TestHdfs: Starting server
...
        --- PASS: TestIntegration/FsMkdir/FsRootCollapse (0.00s)
PASS
ok  	github.com/rclone/rclone/backend/hdfs	11.646s
```
which errors you got ?

**Comment by urykhy** [comment]
some updates:
* switch to new version of client library
* improve integration test

**Comment by ivandeex** [comment]
I recall recently describing all possible steps (you _don't_ need all of them) in https://github.com/rclone/rclone/pull/4075#issuecomment-729060366.

`cd backend/yourbackend; go test -v` is a unit test in rclone parlance.
Integration test is a more thorough thing described in https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request
Could you please run `test_all -remotes TestHdfs: ...`. If it passes, I'll rebase on master, let @ncw know and we're probably ready to merge.

**Comment by ivandeex** [comment]
Please append a _hdfs_ bullet in https://github.com/rclone/rclone/blob/v1.53.3/fstest/test_all/config.yaml

**Comment by urykhy** [comment]
i run `test_all -backends hdfs`

one test fail
```
2020/11/25 22:54:21 FAIL: 1 tests failed in 1m32.958544888s
2020/11/25 22:54:21   * ./sync.test -test.v -test.timeout 1h0m0s -remote TestHdfs: -test.run '^TestSyncUTFNorm$'
2020/11/25 22:54:21     * Failed tests: [TestSyncUTFNorm]
...
    fstest.go:303: 
        	Error Trace:	fstest.go:303
        	            				fstest.go:336
        	            				fstest.go:348
        	            				sync_test.go:1862
        	Error:      	Should be true
        	Test:       	TestSyncUTFNorm
        	Messages:   	listing wrong, want
        	            	  Testêé (14) got
        	            	  Testêé (14), Testêé (18)
    fstest.go:189: 
        	Error Trace:	fstest.go:189
        	            				fstest.go:306
        	            				fstest.go:336
        	            				fstest.go:348
        	            				sync_test.go:1862
        	Error:      	Should be true
        	Test:       	TestSyncUTFNorm
        	Messages:   	Unexpected file "Testêé"
    fstest.go:148: 
        	Error Trace:	fstest.go:148
        	            				fstest.go:193
        	            				fstest.go:306
        	            				fstest.go:336
        	            				fstest.go:348
        	            				sync_test.go:1862
        	Error:      	Not equal: 
        	            	expected: 14
        	            	actual  : 18
        	Test:       	TestSyncUTFNorm
        	Messages:   	Testêé: size incorrect file=14 vs obj=18
```

should i add normalization into backend ?

**Comment by ivandeex** [comment]
@urykhy My guess is ignore this.

@ncw Only one UTF8 test failed for HDFS. Will you have time for review?

**Comment by ivandeex** [comment]
Rebased on current master

**Comment by ivandeex** [comment]
I found a recent comment by @ncw w.r.t. _TestSyncUTFNorm_ https://github.com/rclone/rclone/pull/3699#discussion_r421846902

> This test is really a **test of the sync code** - it does the unicode normalization and checking itself.
> [snip]
> Later on you said you'd like to leave the encoding along - and that is OK with me, so let's leave it like that for the moment.

**Comment by ivandeex** [comment]
Windows build errors are probably related to https://github.com/golang/go/issues/40795

**Comment by ivandeex** [comment]
hub.com/colinmarc/hdfs/v2
The error
```
Error: ../../../../go/pkg/mod/github.com/colinmarc/hdfs/v2@v2.1.1/error.go:41:10: undefined: syscall.ENOTEMPTY
```
from the _other_os_ build (https://github.com/rclone/rclone/pull/4640/checks?check_run_id=1460521596)
either should be fixed by @colinmarc (and reported by someone like you)
or you can rollback to _v1_ for now (we'll not squash the commit at merge to revert it later).

**Comment by ncw** [comment]
> @urykhy My guess is ignore this.
> 
> @ncw Only one UTF8 test failed for HDFS. Will you have time for review?

One failing test isn't a big deal - it is possible to ignore them in the integration tester.

It would be interesting to work out why that test is failing - I wouldn't have expected that particular test to fail.

That failure will annoy macOS users which often have NFK-D characters rather than the more standard NFK-C.

Note I pushed some breaking changes into master recently, so I suggest you rebase and fix the compile errors. It should be obvious what to do (add in context.Context mostly!).

I've given the code a quick once over - it is looking good :-)

Can you rebase and fix the build then I'll review in depth.

Thank you @urykhy :-)

**Comment by urykhy** [comment]
> Note I pushed some breaking changes into master recently, so I suggest you rebase and fix the compile errors. It should be obvious what to do (add in context.Context mostly!).

It's already rebased. my branch on top of c3ac9319f.

> Error: ../../../../go/pkg/mod/github.com/colinmarc/hdfs/v2@v2.1.1/error.go:41:10: undefined: syscall.ENOTEMPTY

https://github.com/colinmarc/hdfs/issues/255

**Comment by ivandeex** [comment]
> Note I pushed some breaking changes into master recently [...snip...] (add in context.Context mostly!).

@urykhy 
Nick means that configuration will be delivered to backends via context in 1.54
HDFS is not on master yet and should pick the new mechanism from the very start.
Please apply a few changes similar to the diff below
Sorry about inconvenience

https://github.com/rclone/rclone/commit/2e21c58e6a6c69be8c2c46cda157d977388687c5#diff-5db312cfcd7e15ce3971d46e6d72450bf10e86a9101904fe3416a21c0410d266R265

**Comment by urykhy** [comment]
@ivandeex i see.
rebased. fix is separate commit. i will squash if it's ok.

**Comment by ivandeex** [comment]
We can squash later at merge.

File: backend/hdfs/fs.go
Comment (ivandeex) [inline backend/hdfs/fs.go:63]: Nick will ask you to remove dead code

Comment (ivandeex) [reply backend/hdfs/fs.go]: Fixed by commit [51f23ac](https://github.com/rclone/rclone/pull/4640/commits/51f23ac1bb97c894c985a8e6b1bc8cd89ef03b36)

**Comment by ivandeex** [comment]
Removed some dead code and pushed two temporary build fixes:
-  fix plan9 build until https://github.com/colinmarc/hdfs/issues/255 is fixed and/or https://github.com/colinmarc/hdfs/pull/256 is accepted
-  fix windows build (like https://github.com/grafana/grafana/pull/28557) until go 1.15.6+ fixes https://github.com/golang/go/issues/40795

@urykhy 
Please save your unfinished commits and pull this down your workdir

In the meantime you can address integration test failure mentioned in https://github.com/rclone/rclone/pull/4640#issuecomment-734494868

File: docs/content/hdfs.md
Comment (ivandeex) [inline docs/content/hdfs.md:121]: Please change source code path to `backend/hdfs/hdfs.go`, rerun `make backenddocs`, verify changes and commit.

File: docs/content/hdfs.md
Comment (ivandeex) [inline docs/content/hdfs.md:12]: This manual page would hreatly benefit from a minimal stripped-down (no multiple OSes, no yarn or sparc examples) _hadoop getting started_, e.g. something like below:
```
sudo apt-get update # (or yum)
sudo apt-get install java
Please go to cloudera site and download yada-yada
tar xzf cloudera-standalone-installer.tgz
run installer...
blah...
configure rclone...
rclone copy ./data hdfs:/newdir1
hadoop fs -list newdir1
```

**Review by ivandeex — CHANGES_REQUESTED** [review]

**Comment by ivandeex** [comment]
Please also add a link from `index.md` to your manual page like [here](https://github.com/rclone/rclone/pull/4566/commits/4b71a1d9101c0fb55f15c8cac4fb458fe4c85904)

File: docs/content/hdfs.md
Comment (urykhy) [reply docs/content/hdfs.md]: done

File: docs/content/hdfs.md
Comment (urykhy) [reply docs/content/hdfs.md:12]: in `fstest/testserver/images/test-hdfs` there is image for tests. is it ok to note about?

maybe with link to [SingleCluster](https://hadoop.apache.org/docs/stable/hadoop-project-dist/hadoop-common/SingleCluster.html)

File: docs/content/hdfs.md
Comment (urykhy) [reply docs/content/hdfs.md]: btw, index.md updated too.

File: docs/content/hdfs.md
Comment (ivandeex) [reply docs/content/hdfs.md:12]: Yes, ok. If the image is present on docker hub then `docker run opts... repo/image:tag cmd...` is enough. If you provide only _Dockerfile_ then add either `docker build ...` or `docker-compose.yml` with a _build_ stanza and link it from docs.
While simply linking to cloudera _getting started_ would be _fine with me too_, I greatly appreciate your idea. A release on external vendor page will change with time while your documented image is under your control. It will give a basis to compare against in new bug/feature requests.

Comment (urykhy) [reply docs/content/hdfs.md:12]: updated.

File: docs/content/hdfs.md
Comment (ivandeex) [inline docs/content/hdfs.md:97]: Normally users don't have rclone source installed.
Let's either prepend
```
git clone https://github.com/rclone/rclone.git
cd rclone
```
or let's ask @ncw to push `rclone/test-hdfs` up https://hub.docker.com/u/rclone and remove build step from docs.

Comment (urykhy) [reply docs/content/hdfs.md]: updated.

File: docs/content/hdfs.md
Comment (ivandeex) [inline docs/content/hdfs.md:133]: Normally rclone backends with filename charset restrictions make use of encoding helpers (e.g. see https://github.com/rclone/rclone/blob/v1.53.3/backend/local/encode_windows.go or https://github.com/rclone/rclone/blob/v1.53.3/backend/local/local.go#L288 etc)
I can't find anything like that here. Do I miss something? Probably it's related to the failing _UTFNorm_ test?

Comment (urykhy) [reply docs/content/hdfs.md:133]: `encoder.MultiEncoder` in `Options`.

**Review by ivandeex — APPROVED** [review]

**Comment by ivandeex** [comment]
@ncw 
This PR will trigger a new build issue described in https://github.com/rclone/rclone/pull/4640#issuecomment-734446145
and currently mitigated by fix https://github.com/rclone/rclone/pull/4640/commits/ec5e8543e260d45ae07d707b99b7377a8676b069
We can either wait for go 1.15.next or factor the fix into separate issue/PR.

**Comment by ivandeex** [comment]
@urykhy 
Does go-HDFS provide information about _data locality_, i.e. is there an API to tell a file stored on the same node as caller/rclone from externally stored one?
I don't request to implement that by this PR, just curious.

**Comment by urykhy** [comment]
> @urykhy
> Does go-HDFS provide information about _data locality_, i.e. is there an API to tell a file stored on the same node as caller/rclone from externally stored one?

it seems so:
* `func (fi *FileInfo) Sys()` return raw `HdfsFileStatusProto` message. 
* this message can contain `LocatedBlocksProto locations`
* `locations.blocks.locs.id` contain `ipAddr` and `hostName`.

**Comment by ivandeex** [comment]
This PR is waiting for @colinmarc reaction on https://github.com/colinmarc/hdfs/pull/256

**Comment by ncw** [comment]
Apologies for the delay in responding. I'd like to merge this for 1.54 if we can :-)

I've looked through the code and I think it is great :-) The integration tests pass (very nearly), the docker image looks great - well done @urykhy  

Thanks for your help shepherding this in @ivandeex 

A single integration test failed - if we can't fix it then mark it as ignore in `fstest/test_all/config.yaml` and that is fine!

image

I note that some of the docs need updating - from [the docs](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#writing-a-new-backend)

These look like the missing ones

- README.md - main GitHub page
- docs/content/overview.md - overview docs
- docs/content/docs.md - list of remotes in config section
- docs/layouts/chrome/navbar.html - add it to the website navigation
- bin/make_manual.py - add the page to the docs constant

### Summary

Things which need to be done before merging

- rebase 
- squash remaining commits
- decide on fate of f08913c12f80a17cfede4670e0e6972c65d67155 - what we could do is exclude plan9 from the hdfs build with build constraints like we do in the [sftp backend](https://github.com/rclone/rclone/blob/f7fe1d766b686078fb429710852a0cf6bdd08179/backend/sftp/sftp.go#L3) note that if we do that then we'll need an [hdfs_unsupported.go](https://github.com/rclone/rclone/blob/master/backend/sftp/sftp_unsupported.go) also. That is probably my preferred approach so we can get this merged.
- decide on fate of ec5e8543e260d45ae07d707b99b7377a8676b069 - are we waiting for a go point release to fix this? which builds did it fail? Maybe we should merge it now and revert it when the point release fixes come in?
- ignore failing integration test 
- add missing docs (these are mostly just one line changes adding HDFS to existing tables)

I think we are very nearly there :-) Thank you for your hard work.

File: backend/all/all.go
Comment (ivandeex) [inline backend/all/all.go:21]: Author of upstream library refused to fix the build for plan9 - https://github.com/colinmarc/hdfs/pull/256#issuecomment-751738114.
I suggest to move this line into a new `all/all_but_plan9.go` with stanza `// +build !plan9`

Comment (ivandeex) [reply backend/all/all.go:21]: But the below fix suggested by @ncw looks better

> like we do in the [sftp backend](https://github.com/rclone/rclone/blob/f7fe1d766b686078fb429710852a0cf6bdd08179/backend/sftp/sftp.go#L3) note that if we do that then we'll need an [hdfs_unsupported.go](https://github.com/rclone/rclone/blob/master/backend/sftp/sftp_unsupported.go) also.

**Comment by RafalSkolasinski** [comment]
+1

**Comment by axsaucedo** [comment]
+1 - this would be extremely useful 👍 is there anything needed to merge @urykhy @ivandeex ?

**Comment by ncw** [comment]
@urykhy and @ivandeex I'm happy to fix up the remaining things if you haven't started already!

**Comment by axsaucedo** [comment]
Awesome, seems it has been approved, thanks for the speedy response - this will be quite useful!

**Comment by urykhy** [comment]
> I'm happy to fix up the remaining things if you haven't started already!

this will be great! 
i will be afk few days.

**Comment by ncw** [comment]
OK all done!

I'll push the result up now.

Note that I kept the buildmode fix @ivandeex since that fix hasn't made it into go1.15.6 according to the git history. We can always revert it later!

If it all goes green then I'll merge it :-)

**Review by ncw — APPROVED** [review]

**Comment by ncw** [comment]
Yay! All tests passed :-)

I'm going to merge this now for v1.54

Thank you very much @urykhy for doing this and @ivandeex for your help.

Anyone following along - I'll post a link in #42 and you can try it and write comments there.

**Comment by miguelpuyol** [comment]
I tested it in a kubernetes cluster to upload data to another S3 remote and it works like a charm! Thank you all for the great tool!

**Comment by ivandeex** [comment]
@ncw Thank you for helping out.

One note regarding the windows build. https://github.com/golang/go/commit/e463c28cc116fb1f40a4e203bddf93b6ef52c8d9 is tagged for `go1.16beta`.
That means we will stay with the build fix until go 1.15 gets phased out in the rclone test suite, unless google decides to backport.
Currently the oldest go in our test suite is 1.11.

**Comment by ncw** [comment]
> One note regarding the windows build. [golang/go@e463c28](https://github.com/golang/go/commit/e463c28cc116fb1f40a4e203bddf93b6ef52c8d9) is tagged for `go1.16beta`.
> That means we will stay with the build fix until go 1.15 gets phased out in the rclone test suite, unless google decides to backport.

That sounds OK. I put a note on the issue to see if they wanted to backport it to go1.15.x

> Currently the oldest go in our test suite is 1.11.

I plan to remove that version shortly so we are only supporting 4 versions then add go1.16beta as a test target.

If by any chance go1.16 is released before rclone 1.54 then we will go with that!

**Pull request merged**
